pkg/lib/bundlev2/synthesis.go

```diff
@@ -0,0 +1,715 @@
+package bundlev2
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	cv1 "k8s.io/api/core/v1"
+	rv1 "k8s.io/api/rbac/v1"
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/kube-aggregator/pkg/apis/apiregistration"
+
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+)
+
+const (
+	DefaultPermission = 0644
+)
+
+var ObjetctTypeNotFoundInBundle = errors.New("object kind not found in bundle")
+
+type gvkn struct {
+	group   string
+	version string
+	kind    string
+	name    string
+}
+
+type bundlev2 struct {
+	bundleDir                   string
+	registryV2CSV               *RegistryV2CSV
+	structuredObjectMap         map[string]*[]unstructured.Unstructured // map[kind]: *array of unstructured objects
+	existingCSVPath             string
+	synthesisedCSV              *v1alpha1.ClusterServiceVersion
+	apiServiceDescriptionMap    map[gvkn]*v1alpha1.APIServiceDescription
+	customResourceDefinitionMap map[gvkn]*v1alpha1.CRDDescription
+	logger                      *log.Entry
+}
+
+// NewBundleV2FromDirectory
+func NewBundleV2FromDirectory(bundleDir string) (*bundlev2, error) {
+	_, err := ioutil.ReadDir(bundleDir)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
I would've expected the body here to populate the object from the directory passed in

Reply:
Well, i usually want to load the resource from the directory here, but last time I think it was Kevin said it would be hard to mock it. I actually not sure what that means. I didn't get the chance to ask. And onward I started to load everything later. in this case, the synthesize function.